src/Components/Address/utils.ts

```diff
@@ -0,0 +1,84 @@
+import { validatePhoneNumber } from "Components/PhoneNumberInput"
+import * as Yup from "yup"
+
+export interface AddressType {
+  addressLine1: string
+  addressLine2: string | null
+  addressLine3: string | null
+  city: string
+  country: string
+  name: string | null
+  phoneNumber: string | null
+  phoneNumberCountryCode: string | null
+  postalCode: string | null
+  region: string | null
+}
+
+export interface SavedAddressType extends AddressType {
+  id: string
+  internalID: string
+  isDefault: boolean
+}
+
+export const EMPTY_ADDRESS = {
+  name: "",
+  country: "US",
+  addressLine1: "",
+  addressLine2: "",
+  addressLine3: "",
+  city: "",
+  phoneNumber: "",
+  phoneNumberCountryCode: "us",
+  postalCode: "",
+  region: "",
+} as const
+
+export const yupPhoneValidator = Yup.string()
+  .required("Phone Number is required")
+  .test({
+    name: "phone-number-is-valid",
+    message: "Please enter a valid phone number",
+    test: (national, context) => {
+      return validatePhoneNumber({
+        national: `${national}`,
+        regionCode: `${context.parent.phoneNumberCountryCode}`,
+      })
+    },
+  })
+
+const usPostalCodeRegexp = /^\d{5}(-\d{4})?$/
+const intlPostalCodeRegexp = /^([A-Z][0-9][A-Z])\s*([0-9][A-Z][0-9])$/
+
+export const postalCodeValidator = Yup.string().when("country", {
+  is: country => country === "US",
+  then: Yup.string()
+    .required("Postal code is required")
+    .matches(usPostalCodeRegexp, "Invalid postal code"),
+  otherwise: Yup.string()
+    .required("Postal code is required")
+    .matches(intlPostalCodeRegexp, "Invalid postal code"),
+})
+
+export const getYupAddressSchema = ({
+  simplePhoneValidation = false,
+} = {}): Yup.ObjectSchema<any, any> => {
+  const phoneNumberRegex = /^[+\-\d]*$/g
+  const phoneValidator = simplePhoneValidation
+    ? Yup.string()
+        .required("Phone number is required")
+        .matches(phoneNumberRegex, "Invalid phone number")
+    : yupPhoneValidator
+
+  return Yup.object().shape({
+    name: Yup.string().required("Name is required"),
+    country: Yup.string().required("Country is required"),
+    addressLine1: Yup.string().required("Address is required"),
+    city: Yup.string().required("City is required"),
+    region: Yup.string().required("Region is required"),
+    postalCode: postalCodeValidator,
+    phoneNumber: phoneValidator,
+    phoneNumberCountryCode: Yup.string().required(
+      "Phone Number Country Code is required"
+    ),
+  })
+}
```

Comment:
Todo: Ended up not using this bit yet, but it might be worth keeping around as we are going to go back to simple phone validation for this first refactor to avoid scope creep. It does have (WIP) unit tests tests which although they are really just testing we use yup correctly could be useful to isolate from the rest of our address form logic. Maybe they would just test the yup schema directly though - depends if we want to keep this factory method for incremental adoption of the better phone input.

Slightly related comment: I noticed that the phone validation query fires on every address change in any field, can we limit this?